src/lib/OS/SimOS.cc

```diff
@@ -0,0 +1,227 @@
+   * be tested to see if it should perform a context switch to the waiting
+   * process. If the core has been executing for longer than execTicks then an
+   * interupt signal is sent to test if a context switch can be made.
+   *
+   * 3. If the interrupt signal is successful, then the head of waitingProcs
+   * queue is put into scheduled queue.
+   *
+   * 4. In order to schedule a process onto a core, a core must be in an Idle
+   * state. An Idle core can only be scheduled a process from the scheduledProcs
+   * queue, i.e. a process which has previously triggered a successful interupt
+   * signal.
+   *
+   * 5. When a process is successfully scheduled, it is removed from the
+   * scheduledProcs queue.
+   *
+   * 6. When a process is de-scheduled from a core it goes back into a waiting
+   * state and is pushed to the back of the waitingProcs queue.
+   *
+   * By using two queues, we ensure that a single process will not interupt
+   * multiple cores. This is due to :
+   *    a. Only a process in the waitingProc queue can cause an interupt
+   *    b. On a successful interupt, the process that caused it is moved to the
+   * scheduledProc queue
+   *    c. Only processes in the scheduledProc queue can be scheduled onto a
+   * core
+   *
+   * If not for this process, then a waitingProc could tell more executing cores
+   * to context switch given a core is likely to be in a switching state (post
+   * successful interupt, pre Idle state) for multiple cycles.
+   * */
+
+  // Loop over all cores, apply correct behaviour based on state
+  for (auto core : cores_) {
+    switch (core->getStatus()) {
+      case CoreStatus::halted: {
+        // Core has had fatal fault. Change SimOS status and return
+        halted_ = true;
+        return;
+      }
+      case CoreStatus::idle: {
+        // Core is idle, schedule head of scheduledProc queue
+        if (!scheduledProcs_.empty()) {
+          // Get context of process that was executing on core
+          OS::cpuContext prevContext = core->getPrevContext();
+          // Core's stored TID will equal -1 if no process has been previously
+          // scheduled (i.e. on first tick of simulation)
+          if (prevContext.TID != -1) {
+            // Find the corresponding process in map
+            auto proc = processes_.find(prevContext.TID)->second;
+            assert(
+                (proc->status_ == procStatus::executing) &&
+                "[SimEng:SimOS] Process updated when not in executing state.");
+            // Only update values which have changed
+            proc->context_.pc = prevContext.pc;
+            proc->context_.regFile = prevContext.regFile;
+            // Change status from Executing to Waiting
+            proc->status_ = procStatus::waiting;
+            waitingProcs_.push(proc);
+          }
+
+          // Schedule new process on core
+          core->schedule(scheduledProcs_.front()->context_);
```

Comment:
As discussed in our call it will be worth checking if a shared_ptr to context can be passed to the core. This way the "state" always remains coherent with the context of the Process. It will also help performance.

Reply:
I will look into this

Reply:
I think given we are unable to fully test the scheduling logic without multi-thread performance, I will look into this after multi-thread support has been implemented

Reply:
Okay sounds good.